### ISSUE TITLE
[NEWS] Na początku 2024 roku wyłączymy zasoby /sale/offers służące do tworzenia i edycji ofert / At the beginning of 2024, we will disable the /sale/offers resources for creating and editing offers

### ISSUE THREAD

**Issue by AureliuszBrussy** [allegro/allegro-api#5729]
Planujemy całkowicie wyłączyć zasoby do tworzenia oraz edycji ofert ze ścieżek **/sale/offers** oraz **/sale/offers/{offerId}**, dlatego:
- **1 stycznia 2023 roku** oznaczymy je jako **Deprecated** i zakończymy ich dalsze wsparcie i rozwój,
- **1 stycznia 2024 roku** całkowicie je **usuniemy**.

W styczniu tego roku udostępniliśmy zasoby **/sale/product-offers** w wersji public, dzięki którym:
- [za pomocą jednego żądania wystawisz aktywną ofertę powiązaną z produktem](https://developer.allegro.pl/tutorials/jak-jednym-requestem-wystawic-oferte-powiazana-z-produktem-D7Kj9gw4xFA),
- [edytujesz dowolne pole oferty](https://developer.allegro.pl/tutorials/jak-jednym-requestem-wystawic-oferte-powiazana-z-produktem-D7Kj9gw4xFA#jak-edytowac-oferte),
- [utworzysz zestaw produktowy](https://developer.allegro.pl/tutorials/jak-jednym-requestem-wystawic-oferte-powiazana-z-produktem-D7Kj9gw4xFA#jak-utworzyc-zestaw-produktowy).

Zasoby **/sale/product-offers** będą jedynymi, które umożliwią wystawienie oraz edycję oferty. Zadbaj o to, żeby przed całkowitym wyłączeniem zasobów **/sale/offers** Twoje oprogramowanie korzystało z odpowiedników tych zasobów na ścieżce **/sale/product-offers**: 

| **Endpoint, który usuniemy** | **Odpowiednik endpointu** | 
| -------- | -------- | 
| [POST /sale/offers](https://developer.allegro.pl/documentation#operation/createOfferUsingPOST) | [POST /sale/product-offers](https://developer.allegro.pl/documentation#operation/createProductOffers) | 
| [PUT /sale/offers/{offerId}](https://developer.allegro.pl/documentation#operation/updateOfferUsingPUT) | [PATCH /sale/product-offers/{offerId}](https://developer.allegro.pl/documentation#operation/editProductOffers) | 
| [GET /sale/offers/{offerId}](https://developer.allegro.pl/documentation#operation/getOfferUsingGET) | [GET /sale/product-offers/{offerId}](https://developer.allegro.pl/documentation#operation/getProductOffer) |

Cały czas pracujemy nad rozwijaniem zasobów **/sale/product-offers**. Weryfikujemy oraz uzupełniamy wszelkie różnice względem **/sale/offers**, bądź wprowadzamy nowe zasoby, które zapewnią ciągłość dostępu do funkcjonalności. 

Przykładem drugiego z tych działań są opcje promowania ofert, które były dostępne w zasobach **/sale/offers**, ale nie pojawiają się w **/sale/product-offers**. Zostały one zastąpione [**dedykowanymi zasobami**](https://developer.allegro.pl/tutorials/jak-zarzadzac-ofertami-7GzB2L37ase#jak-zarzadzac-promowaniem), które dają użytkownikowi dużo większy zakres funkcjonalny do zarządzania opcjami promowania. 

**Dlaczego decydujemy się na wyłączenie zasobów?**

Od początku 2019 roku grupujemy oferty z tym samym produktem, dzięki czemu kupujący łatwiej znajdzie oferty, które spełniają jego oczekiwania. Zwiększamy stopniowo wymagany udział ofert, które są połączone z Katalogiem produktów w wybranych kategoriach, a docelowo wprowadzimy ten wymóg dla wszystkich ofert.

Zasoby **/sale/product-offers** stworzyliśmy z myślą o tym stanie docelowym - wprowadziliśmy w nich istotne usprawnienia przy wystawianiu i edycji ofert powiązanych z produktem. Dlatego zdecydowaliśmy, że będziemy utrzymywać i rozwijać jeden spójny interfejs do zarządzania ofertami z produktem. 

Więcej informacji o zasobach **/sale/product-offers** znajdziesz w [naszym poradniku](https://developer.allegro.pl/tutorials/jak-jednym-requestem-wystawic-oferte-powiazana-z-produktem-D7Kj9gw4xFA).

--------

We plan to completely disable resources for creating and editing offers from the **/sale/offers** and **/sale/offers/{offerId}** paths, therefore:
- on **January 1, 2023**, we will mark them as **Deprecated** and end their further support and development,
- on **January 1, 2024** we will **remove** them completely.

In January this year, we have released the resources **/sale/product-offers** in the public version, which allows you to:
- [list an offer assigned to a product with one request](https://developer.allegro.pl/product_offers/),
- [edit any field of the offer](https://developer.allegro.pl/tutorials/list-offer-assigned-product-one-request-D7Kj9M71Bu6#how-to-edit-an-offer),
- [create a product set](https://developer.allegro.pl/tutorials/list-offer-assigned-product-one-request-D7Kj9M71Bu6#how-to-create-a-product-set).

The **/sale/product-offers** resource will be the only one to allow listing and editing offers. Ensure that your software uses the equivalent of these resources on the **/sale/product-offers** path before we completely disable the **/sale/offers** resource:

| **Endpoint that we will remove** | **Equivalent to endpoint** | 
| -------- | -------- | 
| [POST /sale/offers](https://developer.allegro.pl/documentation#operation/createOfferUsingPOST) | [POST /sale/product-offers](https://developer.allegro.pl/documentation#operation/createProductOffers) | 
| [PUT /sale/offers/{offerId}](https://developer.allegro.pl/documentation#operation/updateOfferUsingPUT) | [PATCH /sale/product-offers/{offerId}](https://developer.allegro.pl/documentation#operation/editProductOffers) | 
| [GET /sale/offers/{offerId}](https://developer.allegro.pl/documentation#operation/getOfferUsingGET) | [GET /sale/product-offers/{offerId}](https://developer.allegro.pl/documentation#operation/getProductOffer) |

We are constantly working on developing **/sale/product-offers** resources. We verify and complete all differences in relation to **/sale/offers**, or introduce new resources that will ensure continuity of access to functionality. 

An example of the second of these actions are the offer promotion options that were available in the **/sale/offers** resource, but do not appear in **/sale/product-offers**. These have been replaced with [**dedicated resources**](https://developer.allegro.pl/tutorials/how-to-process-list-of-offers-m09BKA5v8H3#how-to-manage-promo-options) that give the user a much greater functional coverage to manage promotion options. 

**Why do we choose to turn off resources?**

From the beginning of 2019, we group offers with the same product, making it easier for buyers to find offers that meet their needs. We are gradually increasing the required share of offers that are linked to the Product Catalog in selected categories, and ultimately we will introduce this requirement for all offers.

We created the **/sale/product-offers** resource with this target state in mind - we have introduced significant improvements in listing and editing offers assigned to the product. That is why we decided to maintain and develop one consistent interface for managing offers with the product.

For more information about **/sale/product-offers** resources, see [our guide](https://developer.allegro.pl/tutorials/list-offer-assigned-product-one-request-D7Kj9M71Bu6).

**Comment by tomasz-grzesik** [comment]
Dzięki za wczesne info.

**Comment by niksfirefly** [comment]
Dlaczego
GET https://api.allegro.pl/sale/product-offers
zwraca 405 "The HTTP method is not allowed"
jak moge wylistowac oferty?

**Comment by AureliuszBrussy** [comment]
@niksfirefly aby wylistować oferty korzystaj z [GET /sale/offers](https://developer.allegro.pl/documentation#operation/searchOffersUsingGET). Tego endpointu nie usuniemy. 

Z GETów usuniemy tylko endpoint do pobierania szczegółów pojedynczej oferty, czyli [GET /sale/offers/{offerId}](https://developer.allegro.pl/documentation#operation/getOfferUsingGET) - tutaj dostępna jest alternatywa w postaci [GET /sale/product-offers/{offerId}](https://developer.allegro.pl/documentation#operation/getProductOffer).

**Comment by Radeq** [comment]
@AureliuszBrussy stwierdzenie "edytujesz dowolne pole oferty" w zasobie `/sale/product-offers` nie jest prawdą. Brakuje możliwości ustawienia własnych parametrów, co zostało wspomniane w #5780 . Kiedy mogę się spodziewać aktualizacji?

**Comment by AureliuszBrussy** [comment]
@Radeq zgadza się, mamy to na uwadze - przed wyłączeniem zasobów **/sale/offers** wdrożymy taką możliwość. Oczywiście zakomunikujemy zmianę.

**Comment by pgasidlo** [comment]
@AureliuszBrussy Proszę też pamiętać o bloku z wyróżnieniem, tj. 
```
["promotion"]=>
  object(stdClass)#72 (5) {
    ["emphasized"]=>
    bool(false)
    ["bold"]=>
    bool(false)
    ["highlight"]=>
    bool(false)
    ["departmentPage"]=>
    bool(false)
    ["emphasizedHighlightBoldPackage"]=>
    bool(false)
}
```
Tego też brakuje w nowych metodach.

**Comment by AureliuszBrussy** [comment]
@pgasidlo tak jak wspomnieliśmy w newsie - opcje promowania ofert, które są dostępne w zasobach **/sale/offers**, ale nie pojawiają się w **/sale/product-offers**. Zostały one zastąpione [dedykowanymi zasobami](https://developer.allegro.pl/tutorials/jak-zarzadzac-ofertami-7GzB2L37ase#jak-zarzadzac-promowaniem), które dają użytkownikowi dużo większy zakres funkcjonalny do zarządzania opcjami promowania.

**Comment by maniootek** [comment]
Ja właśnie dostałem powiadomienie o planowanej zmianie dlatego sprawdziłem swój kod. Okazuje się, że faktycznie jeden z moich programów używa zasobu /sale/offers/{offerId} do odczytywania identyfikatora kategorii danej oferty bo jest mi potrzebna do użycia w kalkulatorze opłat i właśnie zmieniłem go na "/sale/product-offers/{offerId} ale dostaje błąd:
`{"errors":[{"code":"OfferAccessDeniedException","message":"Forbidden","details":null,"path":"","userMessage":"Access Denied, You don't have permission to access offer."}]}`
Co może być nie tak? Token jaki jest wpisany do aplikacji to token z konta, które nie wystawiło tej oferty ale do tej pory kod działał. Konto tokena jest powiązane z innymi kontami.
Proszę o pomoc

**Comment by AureliuszBrussy** [comment]
@maniootek czy możesz wskazać przykładowy trace-id wywołania, które kończy się tym błędem? Dokładniej to zweryfikuję.

**Comment by maniootek** [comment]
jak to sprawdzić?

**Comment by AureliuszBrussy** [comment]
@maniootek trace-id znajdziesz w nagłówku naszego response'a z błędem.

**Comment by maniootek** [comment]
trace-id: 79a2c6e9518038e0

**Comment by maniootek** [comment]
ja nie korzystam z /sale/offers tylko z /sale/offers/{offerId}

**Comment by AureliuszBrussy** [comment]
@maniootek ten wątek został wspomniany w innym, stąd powyższe powiadomienie - nie było ono skierowane do Ciebie. 

Sprawdziłem Twój przypadek - oferta należy do innego użytkownika, niż token, który używasz w swoim requeście - zweryfikuj to.

**Comment by maniootek** [comment]
@AureliuszBrussy tak, zgadza się, użyłem tokena konta głównego aby odpytać ofertę powiązanego konta, do tej pory GET /sale/offers/{offerId} pozwalało na to. Czyli GET /sale/product-offers/{offerId} już nie będzie na to pozwalać?

**Comment by AureliuszBrussy** [comment]
@maniootek w API nie istnieje funkcjonalność powiązanego konta - nie powinno być w ogóle możliwości pobrania oferty innym tokenem, niż token użytkownika, do którego oferta należy. Masz przykład trace-id gdzie udało ci się pobrać szczegóły oferty z powiązanego konta (za pomocą /sale/offers/{offerId})?

**Comment by maniootek** [comment]
@AureliuszBrussy ja cały czas tak działam, że używam tokena konta X i odpytuje ofertę konta Y, Z ...
Te konta wszystkie mają ten sam NIP może dlatego mam taką możliwość, proszę podaje trace-id:
4f8a57dcf7fcf266

**Comment by AureliuszBrussy** [comment]
@maniootek przepraszam, pomyłka z mojej strony - [można pobrać](https://github.com/allegro/allegro-api/issues/5002) ofertę z konta powiązanego, jednak nie można jej edytować. To prawidłowe zachowanie. 

W przypadku `GET /sale/product-offers/{offerId}` dodaj nagłówek `x-representative-of` z id użytkownika, do którego należy ta pobierana oferta. W ten sposób pobierzesz ofertę z konta powiązanego i uprawnionego.

**Comment by maniootek** [comment]
Nie planujecie przerobić nowej funkcji, żeby działała jak stara?

**Comment by AureliuszBrussy** [comment]
@maniootek przekazałem sugestię do zespołu odpowiedzialnego za ten zasób, natomiast na ten moment nie planujemy tego zmienić. Poinformuję, jeżeli coś się zmieni w tym temacie.

**Comment by maniootek** [comment]
do stycznia 2024 może się uda, trzymam kciuki

**Comment by maniootek** [comment]
Czyli nie udało się tego zmienić?

**Comment by AureliuszBrussy** [comment]
@maniootek nie, taka decyzja nie zapadła. Na ten moment nie planujemy przerobienia tej funkcjonalności.
